File: lib/opendmm/engines/heyzo.rb
require 'cgi'
require 'opendmm/movie'
require 'opendmm/utils/httparty'

module OpenDMM
  module Engine
    module Heyzo
      def self.search(query)
        query = normalize(query)
        return unless query
        Movie.new(query).details
      end

      private

      def self.normalize(query)
        return unless query =~ /heyzo/i
        return unless query =~ /(\d{3,4})/i
        $1.rjust(4, '0')
      end

      module Site
        include HTTParty
        base_uri 'www.heyzo.com'

        def self.movie(query)
          get "/moviepages/#{query}/index.html"
        end
      end

      class Movie < OpenDMM::Movie
        def initialize(query)
          super(query, Site.movie(query))

          @details.code          = "Heyzo #{query}"
          @details.cover_image   = "/contents/3000/#{query}/images/player_thumbnail_450.jpg"
          @details.maker         = 'Heyzo'

          @details.title         = @html.css('#movie > h1').text
          @details.release_date  = @html.css('#movie > div.info-bg.info-bgWide > div > span.release-day + *').text
          @details.actresses     = @html.css('#movie > div.info-bg.info-bgWide > div > span.actor + *').text.split
          @details.label         = @html.css('#movie > div.info-bg.info-bgWide > div > span.label + *').text.remove(/-+/)
          @details.actress_types = @html.css('#movie > div.info-bg.info-bgWide > div > div.actor-type > span').map(&:text)
          @details.tags          = @html.css('#movie > div.info-bg.info-bgWide > div > div.tag_cloud > ul > li').map(&:text)
          @details.description   = @html.css('#movie > div.info-bg.info-bgWide > div > p').text
        end
      end
    end
  end
end

File: heyzo.go
package opendmm

import (
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"sync"

	"github.com/PuerkitoBio/goquery"
	"github.com/deckarep/golang-set"
	"github.com/golang/glog"
	"github.com/junzh0u/httpx"
)

func heyzoSearch(query string, wg *sync.WaitGroup, metach chan MovieMeta) {
	keywords := heyzoGuess(query)
	for keyword := range keywords.Iter() {
		wg.Add(1)
		go func(keyword string) {
			defer wg.Done()
			heyzoSearchKeyword(keyword, metach)
		}(keyword.(string))
	}
}

func heyzoGuess(query string) mapset.Set {
	keywords := mapset.NewSet()
	matched, _ := regexp.Match("(?i)heyzo", []byte(query))
	if !matched {
		return keywords
	}

	re := regexp.MustCompile("\\d{3,4}")
	matches := re.FindAllString(query, -1)
	for _, match := range matches {
		keywords.Add(fmt.Sprintf("%04s", match))
	}
	return keywords
}

func heyzoGuessFull(query string) mapset.Set {
	keywords := mapset.NewSet()
	for keyword := range heyzoGuess(query).Iter() {
		keywords.Add(fmt.Sprintf("Heyzo %s", keyword))
	}
	return keywords
}

func heyzoSearchKeyword(keyword string, metach chan MovieMeta) {
	glog.Info("Keyword: ", keyword)
	urlstr := fmt.Sprintf(
		"http://www.heyzo.com/moviepages/%s/index.html",
		url.QueryEscape(keyword),
	)
	heyzoParse(urlstr, keyword, metach)
}

func heyzoParse(urlstr string, keyword string, metach chan MovieMeta) {
	glog.V(2).Info("Product page: ", urlstr)
	doc, err := newDocument(urlstr, httpx.GetContentInUTF8(http.Get))
	if err != nil {
		glog.V(2).Infof("Error parsing %s: %v", urlstr, err)
		return
	}

	var meta MovieMeta
	meta.Maker = "Heyzo"
	meta.Code = fmt.Sprintf("Heyzo %s", keyword)
	meta.Page = urlstr

	var urlbase *url.URL
	urlbase, err = url.Parse(urlstr)
	if err != nil {
		return
	}
	var urlcover *url.URL
	urlcover, err = urlbase.Parse(
		fmt.Sprintf("/contents/3000/%s/images/player_thumbnail_450.jpg", keyword))
	if err == nil {
		meta.CoverImage = urlcover.String()
	}

	meta.Title = doc.Find("#movie > h1").Text()
	meta.ReleaseDate = doc.Find("tr.table-release-day > td:nth-child(2)").Text()
	meta.Actresses = doc.Find("tr.table-actor > td:nth-child(2) a").Map(
		func(i int, a *goquery.Selection) string {
			return a.Text()
		})
	meta.Series = doc.Find("tr.table-series > td:nth-child(2)").Text()
	meta.ActressTypes = doc.Find("tr.table-actor-type > td:nth-child(2) a").Map(
		func(i int, a *goquery.Selection) string {
			return a.Text()
		})
	meta.Tags = doc.Find("ul.tag-keyword-list li").Map(
		func(i int, li *goquery.Selection) string {
			return li.Text()
		})
	descNode := doc.Find("tr.table-memo p.memo").Nodes
	if len(descNode) > 0 {
		meta.Description = descNode[0].Data
	}

	metach <- meta
}
